pallets/contracts2/src/lib.rs

```diff
@@ -0,0 +1,573 @@
+        perms: Permissions,
+    ) -> DispatchResultWithPostInfo {
+        Self::general_instantiate(
+            origin,
+            endowment,
+            // Compute the base weight of roughly `base_instantiate`.
+            <T as Config>::WeightInfo::instantiate_with_code(
+                (code.len() / 1024) as u32,
+                (salt.len() / 1024) as u32,
+            ),
+            gas_limit,
+            T::Hashing::hash(&code),
+            Code::Upload(Bytes(code)),
+            inst_data,
+            salt,
+            perms,
+        )
+    }
+
+    /// Instantiates a contract using an existing WASM code blob with `code_hash` as its code.
+    fn base_instantiate_with_hash(
+        origin: T::Origin,
+        endowment: Balance,
+        gas_limit: Weight,
+        code_hash: CodeHash<T>,
+        inst_data: Vec<u8>,
+        salt: Vec<u8>,
+        perms: Permissions,
+    ) -> DispatchResultWithPostInfo {
+        Self::general_instantiate(
+            origin,
+            endowment,
+            // Compute the base weight of roughly `base_instantiate`.
+            <T as Config>::WeightInfo::instantiate_with_hash((salt.len() / 1024) as u32),
+            gas_limit,
+            code_hash,
+            Code::Existing(code_hash),
+            inst_data,
+            salt,
+            perms,
+        )
+    }
+
+    /// General logic for contract instantiation both when the code or code hash is given.
+    ///
+    /// The interesting parameters here are:
+    /// - `base_weight` assigns a base non-variable weight to the full extrinsic.
+    /// - `code_hash` specifies the hash to use to derive the contract's key.
+    /// - `code` specifies the code for the contract, either as a code blob or an existing hash.
+    ///   The hash of `code` in either case is assumed to correspond to `code_hash`.
+    fn general_instantiate(
+        origin: T::Origin,
+        endowment: Balance,
+        base_weight: Weight,
+        gas_limit: Weight,
+        code_hash: CodeHash<T>,
+        code: Code<CodeHash<T>>,
+        inst_data: Vec<u8>,
+        salt: Vec<u8>,
+        perms: Permissions,
+    ) -> DispatchResultWithPostInfo {
+        // Ensure we have perms + we'll need sender & DID.
+        let PermissionedCallOriginData {
+            primary_did: did,
+            sender,
+            ..
+        } = Identity::<T>::ensure_origin_call_permissions(origin)?;
+
+        // Pre-compute what contract's key will be...
+        let contract_key = BaseContracts::<T>::contract_address(&sender, &code_hash, &salt);
+
+        // ...and ensure that key can be a secondary-key of DID...
+        Identity::<T>::ensure_secondary_key_can_be_added(&did, &contract_key, &perms)?;
+
+        with_transaction(|| {
+            // Roll back `join_identity` if contract was not instantiated.
+            // ...so that the CDD due to `endowment` passes.
+            Identity::<T>::unsafe_join_identity(did, perms, contract_key);
```

Comment:
Being able to spawn a new contract from another is a useful feature.  Since contracts can't be upgraded, a proxy contract can be used that to forward calls to the newest contract version.